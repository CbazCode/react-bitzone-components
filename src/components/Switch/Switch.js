import styled, { createGlobalStyle, css } from "styled-components";


//color
//size


//We have id customize !!
//We have disabled  --- 
//We have checked we need an onchange  ---
//We have default checked --
//We have isRequired!! --
//We have name!! --
//We have size !! --
//We have color


const setBgColor = (bgColor = "default")=> {
   
    switch (bgColor) {
        case "red":
            return '#DC2626'
        case "yellow":
            return '#FBBF24'
        case "green":
            return '#34D399'
        case "blue":
            return '#3B82F6'
        case "purple":
            return '#8B5CF6'
        case "indigo":
            return '#6366F1'
        case "pink":
            return '#EC4899'
        default:
            return "#111";
    }
}


const setDefaultCheked = (defaultRight) => {

    return !defaultRight ?  'none' : 'translateX(24px)'
}

const setDefaultChekedBall = (defaultRight) => {
  
    return !defaultRight ?  'translateX(24px)': 'none'
}

const setSizeBall = (size) => {
    switch (size) {
        case "sm":
            return "15px";
        case "md":
            return "22px";
        case "lg":
            return "30px";
        default:
            return "22px";
    
    }
    
}

const setSizeLabel = (size)=>{
    switch (size) {
        case "sm":
            return (
                css`
                    height: 19px;
	                width: 43px;
                `);
        case "md":
            return  (
                css`
                    height: 26px;
	                width: 50px;
                `);
        case "lg":
            return (
                css`
                    height: 34px;
	                width: 58px;
                `);
        default:
            return (
                css`
                    height: 26px;
	                width: 50px;
                `);
    
    }

}

export const SwitchContainer = styled.div`
    .checkbox:checked + .label .ball{
        transform: ${({defaultRight})=> setDefaultChekedBall(defaultRight)}

    }
`

export const InputSwitch = styled.input.attrs((props) => (
    {
        type: "checkbox",
        id:props.id,
        disabled: props.isDisabled,
        checked: props.isChecked,
        required: props.isRequired,
        name: props.name
    }
))
` 
    opacity: 0;


`
export const LabelSwitch = styled.label.attrs((props)=>(
    {
        htmlFor:props.htmlFor
    }
))
`
    background-color: ${({bgColor}) => setBgColor(bgColor)};
	border-radius: 50px;
	cursor: pointer;
	display: flex;
	align-items: center;
	justify-content: space-between;
	padding: 5px;
	position: relative;
	${({size})=> setSizeLabel(size)};
	transform: scale(1.5);
`

export const Ball = styled.div`
    background-color: #fff;
	border-radius: 50%;
	position: absolute;
	top: 2px;
	left: 2px;
	height: ${({size})=> setSizeBall(size)};
	width: ${({size})=> setSizeBall(size)};
	transition: transform 0.2s linear;
    transform: ${({defaultRight})=> setDefaultCheked(defaultRight)};

`






