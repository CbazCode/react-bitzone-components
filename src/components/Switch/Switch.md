```js
<Switch/>
```
```js
<Switch  bgColor="primary" defaultRight  />
```
```js
<Switch id="idTest1" size="xl" bgColor="secondary" />
```
```js
<Switch bgColor="success" />
```
```js
<Switch bgColor="accent" isChecked  />
```