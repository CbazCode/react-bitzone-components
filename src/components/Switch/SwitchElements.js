import styled, { css } from 'styled-components';
import { setBgColor } from '../../utils/helpers';

const setDefaultCheked = (defaultRight) => {
  return !defaultRight ? 'none' : 'translateX(24px)';
};

const setDefaultChekedBall = (defaultRight) => {
  return !defaultRight ? 'translateX(24px)' : 'none';
};

const setSizeBall = (size) => {
  switch (size) {
    case 'sm':
      return '15px';
    case 'md':
      return '22px';
    case 'lg':
      return '30px';
    default:
      return '22px';
  }
};

const setSizeLabel = (size) => {
  switch (size) {
    case 'sm':
      return css`
        height: 19px;
        width: 43px;
      `;
    case 'md':
      return css`
        height: 26px;
        width: 50px;
      `;
    case 'lg':
      return css`
        height: 34px;
        width: 58px;
      `;
    default:
      return css`
        height: 26px;
        width: 50px;
      `;
  }
};

export const SwitchContainer = styled.div`
  margin: 0.5rem;
  box-sizing: border-box;
  .checkbox:checked + .label .ball {
    transform: ${({ defaultRight }) => setDefaultChekedBall(defaultRight)};
  }
`;

export const InputSwitch = styled.input.attrs((props) => ({
  type: 'checkbox',
  id: props.id,
  disabled: props.isDisabled,
  checked: props.isChecked,
  required: props.isRequired,
  name: props.name,
}))`
  position: absolute;
  opacity: 0;
`;
export const LabelSwitch = styled.label.attrs((props) => ({
  htmlFor: props.htmlFor,
}))`
   box-sizing: border-box;
  background-color: ${({ bgColor }) => setBgColor(bgColor)};
  border-radius: 50px;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 5px;
  position: relative;
  ${({ size }) => setSizeLabel(size)};
  transform: scale(1.5);
`;

export const Ball = styled.div`
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 50%;
  position: absolute;
  top: 2px;
  left: 2px;
  height: ${({ size }) => setSizeBall(size)};
  width: ${({ size }) => setSizeBall(size)};
  transition: transform 0.2s linear;
  transform: ${({ defaultRight }) => setDefaultCheked(defaultRight)};
`;
