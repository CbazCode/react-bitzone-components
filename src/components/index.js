export { default as Button } from './Button/Button';
export { default as Drawer } from './Drawer/Drawer';
export { default as Input } from './Input/Input';
export { default as Modal } from './Modal/Modal';
export { default as Skeleton } from './Skeleton/Skeleton';
export { default as Slider } from './Slider/Slider';
export { default as Spinner } from './Spinner/Spinner';
export { default as Switch } from './Switch/Switch';
