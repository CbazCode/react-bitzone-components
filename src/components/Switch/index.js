import {Ball, InputSwitch, LabelSwitch, SwitchContainer} from "./Switch";

const Switch = ({id, isDisabled,isChecked, defaultCheked, isRequired, name, size, bgColor}) => {
    console.log(defaultCheked);
    return (
        <SwitchContainer defaultCheked ={defaultCheked}>
            <InputSwitch className="checkbox" id={id} isDisabled={isDisabled} isChecked ={isChecked} isRequired={isRequired} name={name}/>
            <LabelSwitch className="label" htmlFor={id} size={size} bgColor={bgColor}>
                <Ball className="ball" defaultCheked ={defaultCheked} size={size}/>
            </LabelSwitch>
        </SwitchContainer>  
        
    );
}
 
export default Switch;