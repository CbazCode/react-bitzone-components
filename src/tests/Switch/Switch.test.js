import { Switch } from '../../components';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { InputSwitch } from '../../components/Switch/Switch';

describe('Test en <Switch />', () => {
  test('Debe renderizarse el componente por defecto', () => {
    const { container } = render(<Switch />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('componente debe estar con isChecked ', () => {
    const wrapper = shallow(<Switch isChecked />);
    const attribute = wrapper.getElement('div').props.children[0].props;
    expect(attribute.isChecked).toEqual(true);
  });
  test('componente debe estar con name ', () => {
    const wrapper = shallow(<Switch name="test" />);
    const attribute = wrapper.getElement('div').props.children[0].props;
    expect(attribute.name).toEqual('test');
  });
  test('componente debe estar con id establecido ', () => {
    const wrapper = shallow(<Switch id="testid" />);
    const attribute = wrapper.getElement('div').props.children[0].props;
    expect(attribute.id).toEqual('testid');
  });

  test('componente debe estar con isRequired ', () => {
    const wrapper = shallow(<Switch isRequired />);
    const attribute = wrapper.getElement('div').props.children[0].props;

    expect(attribute.isRequired).toEqual(true);
  });

  test('componente debe deshabilitarse', () => {
    const { container } = render(<InputSwitch isDisabled />);
    expect(container.firstChild).toBeDisabled();
  });

  test('componente debe renderizarse en el tamaÃ±o deseado', () => {
    const wrapper = shallow(<Switch size="sm" />);
    const attribute = wrapper.getElement('div').props.children[1].props;
    expect(attribute.size).toEqual('sm');
  });

  test('componente debe renderizarse con el color deseado', () => {
    const wrapper = shallow(<Switch bgColor="red" />);
    const attribute = wrapper.getElement('div').props.children[1].props;
    expect(attribute.bgColor).toEqual('red');
  });
});
