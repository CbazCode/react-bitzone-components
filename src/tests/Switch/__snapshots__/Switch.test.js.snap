// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test en <Switch /> Debe renderizarse el componente por defecto 1`] = `
.c0 {
  margin: 0.5rem;
  box-sizing: border-box;
}

.c0 .checkbox:checked + .label .ball {
  -webkit-transform: translateX(24px);
  -ms-transform: translateX(24px);
  transform: translateX(24px);
}

.c1 {
  position: absolute;
  opacity: 0;
}

.c2 {
  box-sizing: border-box;
  background-color: #9CA3AF;
  border-radius: 50px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 5px;
  position: relative;
  height: 26px;
  width: 50px;
  -webkit-transform: scale(1.5);
  -ms-transform: scale(1.5);
  transform: scale(1.5);
}

.c3 {
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 50%;
  position: absolute;
  top: 2px;
  left: 2px;
  height: 22px;
  width: 22px;
  -webkit-transition: -webkit-transform 0.2s linear;
  -webkit-transition: transform 0.2s linear;
  transition: transform 0.2s linear;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

<div
  class="c0"
>
  <input
    class="c1 checkbox"
    id="0.5931"
    type="checkbox"
  />
  <label
    class="c2 label"
    for="0.5931"
  >
    <div
      class="c3 ball"
    />
  </label>
</div>
`;
